### DISCUSSION TITLE
Is there a way to schedule a new refetch without cancelling an existing refetch?

### DISCUSSION THREAD

**Discussion by gaborcs** [TanStack/query#8035] [Q&A] [answered]
I'd like to trigger a refetch when I know the relevant data has changed (based on a timestamp update), but it seems problematic when an existing refetch is in progress:
- Canceling the refetch that is already in progress leads to dropping optimistic updates that are applied via the UI, because the mutation will not be in a pending state anymore even though the refetch hasn't completed yet.
- Using `cancelRefetch: false` leads to not triggering a new refetch at all. The existing refetch may return outdated data, so this is not great either.

Is there a way to schedule a new refetch without cancelling the existing refetch?

**Comment by TkDodo** [comment, marked as answer]
> Canceling the refetch that is already in progress leads to dropping optimistic updates that are applied via the UI, because the mutation will not be in a pending state anymore even though the refetch hasn't completed yet.

I guess you are talking about optimistic updates that are done by rendering `variables` in the UI? Because cache-based optimistic updates wouldn't be affected here.

**Reply by gaborcs** [reply]
Yes, I was doing optimistic updates by rendering `variables` in the UI, because I only needed the updates to be reflected in a single component (and its children), and this approach seemed simpler. However, I think cache-based optimistic updates would also have a similar issue. If I trigger a refetch while a mutation is being performed, and the refetch completes before the mutation does, then the cache may be updated to a state that doesn't have the result of the mutation yet, dropping the optimistic update.

**Reply by TkDodo** [reply]
yeah but that's really an edge case and it will be eventually consistent.

**Reply by gaborcs** [reply]
That's true. My fear is that stuff like that tends to cause flakiness in E2E tests, so I'd prefer to avoid such issues if possible. But it does sound like a better approach than what I currently have, so thank you for the tip, I'll probably change our implementation to use the cache-based solution.

I guess doing optimistic updates right is hard, especially when combined with real-time updates. Maybe I should look into using a sync engine like Replicache or Zero at some point.
